core/trino-main/src/main/java/io/trino/cost/StatisticRange.java

```diff
@@ -122,7 +123,17 @@ public double overlapPercentWith(StatisticRange other)
         if (isInfinite(length()) && isFinite(lengthOfIntersect)) {
             return INFINITE_TO_FINITE_RANGE_INTERSECT_OVERLAP_HEURISTIC_FACTOR;
         }
+
         if (lengthOfIntersect > 0) {
+            double thisDensity = this.distinctValues / length();
+            double otherDensity = other.distinctValues / other.length();
+            double minDensity = minExcludeNaN(thisDensity, otherDensity);
+
+            if (!isNaN(thisDensity) && !isNaN(otherDensity)
+                    && isFinite(length()) && isFinite(other.length())
+                    && minDensity < DENSITY_HEURISTIC_THRESHOLD) {
+                return minExcludeNaN(this.distinctValues, other.distinctValues) / this.distinctValues;
```

Comment:
can this be `return min(other.distinctValues / this.distinctValues, 1);` ?

Reply:
I think that we cannot use:
`min(other.distinctValues / this.distinctValues, 1)`
because in cases like `IN (1, 2, 3, 4)`, the distinctValues for other is `NaN`. Then:
`min(other.distinctValues / this.distinctValues, 1) = min(NaN, 1) = NaN`
so the result would be `NaN`, which is not valid.

Instead, we should use:
`minExcludeNaN(this.distinctValues, other.distinctValues) / this.distinctValues`
or equivalently:
`minExcludeNaN(other.distinctValues / this.distinctValues, 1)`

---

Also, I looked more carefully at the idea of removing the `weight` I added before. If we do that, the estimate will go from about `0.29` to `1`, `because other.distinctValues` is `NaN`